File: walkthroughs/python/changing_a_table/table_files.py
# What a snapshot says the table is: its data files and delete files, from the table's metadata.
import json

snapshots = json.load(open("fixtures/changes/_snapshots.json"))["snapshots"]
snapshot = next(s for s in snapshots if s["id"] == "after-a-day")

for f in snapshot["data_files"]:
    per_row = f["file_size"] // f["record_count"]
    print(f"{f['path']}: {f['record_count']} rows in {f['file_size']} bytes, {per_row} bytes a row")
print(len(snapshot["data_files"]), "data files and", len(snapshot["delete_files"]), "delete files")

File: walkthroughs/src/bin/table_files.rs
//! What a snapshot says the table is: its data files and delete files, from the table's metadata.

use parquet_lab::json::Json;

fn main() {
    let text = std::fs::read_to_string("fixtures/changes/_snapshots.json")
        .expect("run this from the repository's root");
    let all = Json::parse(&text).expect("JSON");
    let snapshot = all
        .get("snapshots")
        .and_then(Json::as_array)
        .and_then(|s| {
            s.iter()
                .find(|s| s.get("id").and_then(Json::as_str) == Some("after-a-day"))
        })
        .expect("a snapshot called after-a-day");
    let list = |k: &str| snapshot.get(k).and_then(Json::as_array).unwrap();

    for f in list("data_files") {
        let path = f.get("path").and_then(Json::as_str).unwrap();
        let rows = f.get("record_count").and_then(Json::as_i64).unwrap();
        let size = f.get("file_size").and_then(Json::as_i64).unwrap();
        let per_row = size / rows;
        println!("{path}: {rows} rows in {size} bytes, {per_row} bytes a row");
    }
    println!(
        "{} data files and {} delete files",
        list("data_files").len(),
        list("delete_files").len()
    );
}
